File: old/arch/x86/vga.zig
// file : vga.zig
//
// Copyright (C) 2018  Joshua Barretto <joshua.s.barretto@gmail.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

const char = @import("../../util/char.zig");
const log = @import("../../util/log.zig");
const mmap= @import("mmap.zig");
const cpu = @import("cpu.zig");

pub const Color = enum(u4) {
	BLACK = 0x0,
	BLUE = 0x1,
	GREEN = 0x2,
	CYAN = 0x3,
	RED = 0x4,
	MAGENTA = 0x5,
	BROWN = 0x6,
	GRAY = 0x7,
	DARK_GRAY = 0x8,
	BRIGHT_BLUE = 0x9,
	BRIGHT_GREEN = 0xA,
	BRIGHT_CYAN = 0xB,
	BRIGHT_RED = 0xC,
	BRIGHT_MAGENTA = 0xD,
	YELLOW = 0xE,
	WHITE = 0xF,
};

const Entry = packed struct {
	c: u8,
	fg: Color,
	bg: Color,

	pub fn new(c: u8, fg: Color, bg: Color) Entry {
		return Entry{ .c = c, .fg = fg, .bg = bg };
	}
};

const PORT_CMD: u16 = 0x03D4;
const PORT_DATA: u16 = 0x03D5;

const COLS: u16 = 80;
const ROWS: u16 = 25;
const TAB_WIDTH = 4;

const default_text_color = Color.WHITE;
const default_back_color = Color.BLACK;

const vmem = @intToPtr(&volatile Entry, mmap.MEM_VGA_TEXTMODE)[0..COLS * ROWS];

var cursor_pos: u16 = 0;
var text_color: Color = Color.WHITE;
var back_color: Color = Color.BLACK;

pub fn init() %void {
	// Reset textmode parameters
	cursor_pos = 0;
	text_color = Color.WHITE;
	back_color = Color.BLACK;

	// Clean the video memory
	for (vmem) |*entry| {
		*entry = Entry.new(' ', default_text_color, default_back_color);
	}

	log.bootf(true, "VGA {}x{} mode initiated", COLS, ROWS);
}

fn alignLower(index: u16, mod: u16) u16 {
	return index + mod - index % mod;
}

pub fn writeChar(c: u8) void {
	switch (c) {
			// Newline, skip to next row
		'\n' => cursor_pos = alignLower(cursor_pos + 1, COLS),
			// Tab, skip to next tab alignment
		'\t' => cursor_pos = alignLower(cursor_pos + 1, TAB_WIDTH),
			// Printable, display character
		else => if (char.isPrintable(c)) {
				vmem[cursor_pos] = Entry.new(c, text_color, back_color);
				cursor_pos += 1;
			},
	}

	if (cursor_pos >= COLS * ROWS)
		cursor_pos = 0;

	setCursorPosition(cursor_pos);
}

pub fn writeStr(str: []const u8) void {
	for (str) |c| {
		writeChar(c);
	}
}

pub fn placeCursor(col: u8, row: u8) void {
	cursor_pos = COLS * row + col;
	setCursorPosition(cursor_pos);
}

pub fn getDefaultTextColor() Color {
	return default_text_color;
}

pub fn getDefaultBackColor() Color {
	return default_back_color;
}

pub fn setTextColor(color: Color) void {
	text_color = color;
}

pub fn setBackColor(color: Color) void {
	back_color = color;
}

pub fn setCursorEnabled(comptime enabled: bool) void {
	if (enabled) {
		cpu.out8(PORT_CMD, 0x0A);
		cpu.out8(PORT_DATA, 0x00);
	} else {
		cpu.out8(PORT_CMD, 0x0A);
		cpu.out8(PORT_DATA, 0x3F);
	}
}

pub fn setCursorPosition(offset: u16) void {
	cpu.out8(PORT_CMD, 0x0F);
	cpu.out8(PORT_DATA, u8(offset & 0xFF));
	cpu.out8(PORT_CMD, 0x0E);
	cpu.out8(PORT_DATA, u8((offset >> 8) & 0xFF));
}

File: src/driver/video/vga.rs
// file : vga.rs
//
// Copyright (C) 2018  Joshua Barretto <joshua.s.barretto@gmail.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use core::ptr::Unique;
use volatile::Volatile;
use spin::Mutex;

use llapi::intrinsic::chipset::regions::VGA_TEXTMODE_RAM;

pub const COLS: usize = 80;
pub const ROWS: usize = 25;
pub const TAB_WIDTH: usize = 4;

#[allow(dead_code)]
#[derive(Debug, Copy, Clone)]
#[repr(u8)]
pub enum Color {
	Black      = 0,
	Blue       = 1,
	Green      = 2,
	Cyan       = 3,
	Red        = 4,
	Magenta    = 5,
	Brown      = 6,
	LightGray  = 7,
	DarkGray   = 8,
	LightBlue  = 9,
	LightGreen = 10,
	LightCyan  = 11,
	LightRed   = 12,
	Pink       = 13,
	Yellow     = 14,
	White      = 15,
}

#[derive(Debug, Copy, Clone)]
#[repr(C, packed)]
struct Entry {
	c: u8,
	fmt: u8,
}

pub struct Writer {
	cursor: usize,
	fg_color: Color,
	bg_color: Color,
	buffer: Unique<[Volatile<Entry>; COLS * ROWS]>,
}

// TODO: Work out why static initialization doesn't work
static WRITER: Mutex<Writer> = Mutex::new(Writer {
	cursor: 0,
	fg_color: Color::White,
	bg_color: Color::Black,
	buffer: unsafe { Unique::new_unchecked(0 as *mut _) },
});

extern {
	fn _vga_boot_cursor() -> usize;
}

fn colors_to_fmt(fg: Color, bg: Color) -> u8 {
	((bg as u8) << 4) | fg as u8
}

impl Entry {
	fn empty(fg: Color, bg: Color) -> Entry {
		Entry {
			c: b' ',
			fmt: colors_to_fmt(fg, bg),
		}
	}
}

impl Writer {
	fn init(&mut self) {
		self.cursor = unsafe { _vga_boot_cursor() };
		self.fg_color = Color::White;
		self.bg_color = Color::Black;
		self.buffer = unsafe { Unique::new_unchecked(VGA_TEXTMODE_RAM as *mut _) };
	}

	fn write(&mut self, c: u8) {
		match c {
			b'\n' => self.cursor += COLS - (self.cursor % COLS),
			b'\t' => self.cursor += TAB_WIDTH - (self.cursor % TAB_WIDTH),
			c => {
				let cursor = self.cursor;
				let fmt = colors_to_fmt(self.fg_color, self.bg_color);
				self.buffer()[cursor].write(Entry {
					c: c,
					fmt: fmt,
				});
				self.cursor += 1;
			}
		};

		while self.cursor >= COLS * ROWS {
			self.scroll(1);
			self.cursor -= COLS
		}
	}

	fn scroll(&mut self, lines: usize) {
		let chars = lines * COLS;
		for i in 0..COLS * (ROWS - lines) {
			let old = self.buffer()[COLS + i].read();
			self.buffer()[i].write(old);
		}
		for i in 0..chars {
			let blank = Entry::empty(self.fg_color, self.bg_color);
			self.buffer()[COLS * ROWS - (i + 1)].write(blank);
		}
	}

	fn buffer(&mut self) -> &mut [Volatile<Entry>; COLS * ROWS] {
		unsafe { self.buffer.as_mut() }
	}
}

use spin::Once;
static INIT: Once<()> = Once::new();

pub fn init() {
	INIT.call_once(|| {
		WRITER.lock().init();
		logok!("Initiated VGA driver");
	});
}

pub fn write_char(c: char) {
	WRITER.lock().write(c as u8)
}
